crates/bonsaidb-macros/src/lib.rs

```diff
@@ -723,6 +727,134 @@ pub fn api_derive(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+                parse_quote!(::#ident)
+            }
+            Ok(FoundCrate::Itself) => parse_quote!(crate),
+            Err(_) if cfg!(feature = "omnibus-path") => parse_quote!(::bonsaidb::files),
+            Err(_) => parse_quote!(::bonsaidb_core),
+        },
+    }
+}
+
+#[derive(Attribute)]
+#[attribute(ident = "file_config")]
+#[attribute(
+    invalid_field = r#"`metadata = MetadataType`, block_size = 65_536, `authority = "authority"`, `files_name = "files"`, `blocks_name = "blocks"`, `core = bonsaidb::core` and `files = bonsaidb::files` are supported attributes"#
+)]
+struct FileConfigAttribute {
+    #[attribute(expected = r#"Specify the metadata type like so: `metadata = MetadataType`"#)]
+    metadata: Option<Type>,
+    #[attribute(expected = r#"Specify the block size like so: `block_size = 65_536`"#)]
+    block_size: Option<usize>,
+    authority: Option<String>,
+    files_name: Option<String>,
+    blocks_name: Option<String>,
+    #[attribute(expected = r#"Specify the the path to `core` like so: `core = bosaidb::core`"#)]
+    core: Option<Path>,
+    #[attribute(
+        expected = r#"Specify the the path to bonsaidb `files` like so: `files = bosaidb::files`"#
+    )]
+    files: Option<Path>,
+}
+
+/// Derives the `bonsaidb::files::FileConfig` trait.
+#[proc_macro_error]
+/// `#[api(metadata = MetadataType, block_size = 65_536, authority = "authority", files_name = "files", blocks_name = "blocks", core = bonsaidb::core, files = bosaidb::files)]`
+/// all arguments are optional
+#[proc_macro_derive(FileConfig, attributes(file_config))]
+pub fn file_config_derive(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let DeriveInput {
+        attrs,
+        ident,
+        generics,
+        ..
+    } = parse_macro_input!(input as DeriveInput);
+
+    let FileConfigAttribute {
+        metadata,
+        block_size,
+        authority,
+        files_name,
+        blocks_name,
+        core,
+        files,
+    } = FileConfigAttribute::from_attributes(&attrs).unwrap_or_abort();
+
+    let core = core.unwrap_or_else(core_path);
+    let files = files.unwrap_or_else(files_path);
+    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
+    // TODO error if only authority is given
+    // let (files_name, blocks_name) = if let (Some (authority) = authority {
+    //     let files_name = files_name.unwrap_or_else(|| abort_call_site!("if `authority` is given, `files_name` and `blocks_name need to be provided as well"));
+    //     (,())
+    // }
+
+    let (files_name, blocks_name) = match (authority, files_name, blocks_name) {
+        (None, None, None) => (
+            quote!(#files::BonsaiFiles::files_name()),
+            quote!(#files::BonsaiFiles::blocks_name()),
+        ),
+        (Some(authority), Some(files_name), Some(blocks_name)) => (
+            quote!(#core::schema::Qualified::new(#authority, #files_name)),
```

Comment:
Huh, invoking `Qualified::new` works here? I think some of the other code has a more complex `as` expression that could be simplified. That doesn't need to be done in this PR, just so astounded that this works that I'm leaving the comment :)

Neat!